File: website/lib/access-tokens.ts
import fs from 'fs';
import path from 'path';

export interface AccessToken {
  id: string;
  name: string;
  email: string;       // always lowercase
  company: string;
  role: string;
  erpSystem: string;
  message: string;
  submittedAt: string; // ISO
  status: 'pending' | 'active' | 'expired';
  accessToken: string; // empty until approved
  expiresAt: string | null; // ISO, null until approved
}

const DATA_FILE = path.join(process.cwd(), 'data', 'access-tokens.json');

function readAll(): AccessToken[] {
  if (!fs.existsSync(DATA_FILE)) return [];
  try {
    return JSON.parse(fs.readFileSync(DATA_FILE, 'utf-8')) as AccessToken[];
  } catch {
    return [];
  }
}

function writeAll(tokens: AccessToken[]) {
  const dir = path.dirname(DATA_FILE);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(DATA_FILE, JSON.stringify(tokens, null, 2));
}

function hex(bytes = 16): string {
  const arr = new Uint8Array(bytes);
  globalThis.crypto.getRandomValues(arr);
  return Array.from(arr, (b) => b.toString(16).padStart(2, '0')).join('');
}

export function createPendingEntry(
  data: Omit<AccessToken, 'id' | 'status' | 'accessToken' | 'expiresAt'>
): AccessToken {
  const tokens = readAll();
  const entry: AccessToken = {
    ...data,
    id: `req_${Date.now()}_${hex(4)}`,
    status: 'pending',
    accessToken: '',
    expiresAt: null,
  };
  tokens.push(entry);
  writeAll(tokens);
  return entry;
}

export function approveEntry(
  id: string,
  ttlHours: number
): AccessToken | null {
  const tokens = readAll();
  const idx = tokens.findIndex((t) => t.id === id);
  if (idx === -1) return null;

  const accessToken = hex(16); // 32-char hex = 128-bit entropy
  const expiresAt = new Date(Date.now() + ttlHours * 3_600_000).toISOString();
  tokens[idx] = { ...tokens[idx], status: 'active', accessToken, expiresAt };
  writeAll(tokens);
  return tokens[idx];
}

export function findActiveToken(email: string, code: string): AccessToken | null {
  const tokens = readAll();
  return (
    tokens.find(
      (t) =>
        t.email === email.toLowerCase().trim() &&
        t.accessToken === code &&
        t.status === 'active' &&
        t.expiresAt !== null &&
        new Date(t.expiresAt) > new Date()
    ) ?? null
  );
}

export function findById(id: string): AccessToken | null {
  return readAll().find((t) => t.id === id) ?? null;
}

File: backend/access_tokens.py
"""
Access token CRUD — stores pending/active/expired tokens in data/access-tokens.json.
Key names match the TypeScript AccessToken interface for file compatibility.
"""

import json
import os
import pathlib
import secrets
import time
from typing import Optional

_DATA_FILE = pathlib.Path("data/access-tokens.json")


def _read() -> list[dict]:
    if not _DATA_FILE.exists():
        return []
    try:
        return json.loads(_DATA_FILE.read_text(encoding="utf-8"))
    except Exception:
        return []


def _write(tokens: list[dict]) -> None:
    _DATA_FILE.parent.mkdir(parents=True, exist_ok=True)
    _DATA_FILE.write_text(
        json.dumps(tokens, indent=2, ensure_ascii=False), encoding="utf-8"
    )


def _hex(n: int = 16) -> str:
    return secrets.token_hex(n)


def create_pending(
    *,
    name: str,
    email: str,
    company: str,
    role: str,
    erp_system: str,
    message: str,
    submitted_at: str,
) -> dict:
    tokens = _read()
    entry = {
        "id": f"req_{int(time.time() * 1000)}_{_hex(4)}",
        "name": name,
        "email": email.lower().strip(),
        "company": company,
        "role": role,
        "erpSystem": erp_system,
        "message": message,
        "submittedAt": submitted_at,
        "status": "pending",
        "accessToken": "",
        "expiresAt": None,
    }
    tokens.append(entry)
    _write(tokens)
    return entry


def approve(id_: str, ttl_hours: int) -> Optional[dict]:
    tokens = _read()
    idx = next((i for i, t in enumerate(tokens) if t["id"] == id_), None)
    if idx is None:
        return None
    access_token = _hex(16)  # 32-char hex, 128-bit entropy
    expires_ms = int(time.time() * 1000) + ttl_hours * 3_600_000
    import datetime
    expires_at = datetime.datetime.fromtimestamp(
        expires_ms / 1000, tz=datetime.timezone.utc
    ).isoformat()
    tokens[idx] = {
        **tokens[idx],
        "status": "active",
        "accessToken": access_token,
        "expiresAt": expires_at,
    }
    _write(tokens)
    return tokens[idx]


def find_active(email: str, code: str) -> Optional[dict]:
    import datetime
    now = datetime.datetime.now(tz=datetime.timezone.utc)
    for t in _read():
        if (
            t["email"] == email.lower().strip()
            and t["accessToken"] == code
            and t["status"] == "active"
            and t.get("expiresAt")
            and datetime.datetime.fromisoformat(t["expiresAt"]) > now
        ):
            return t
    return None


def find_by_id(id_: str) -> Optional[dict]:
    return next((t for t in _read() if t["id"] == id_), None)
